### PR TITLE
MM-39798: expand channel and team membership

### PR DESCRIPTION
**Pull request by levb**
#### Summary
Added support for Expand.(Channel|Team)Member and test

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-39798

No testing required since it's not used yet

### CODE REVIEW

**Comment by levb** [comment]
@hanzei @catalintomai 1/5 we can file a few HW tickets to add more `Expand` tests, following the pattern here. Should we do it? If so, in GitHub alone, or start in Jira?

**Comment by catalintomai** [comment]
@levb , I'd say start in Jira, mainly for tracking purposes.

File: apps/expand.go [hunk 1]
```diff
@@ -48,6 +48,10 @@ type Expand struct {
 	// DisplayName, Name
 	Channel ExpandLevel `json:"channel,omitempty"`
 
+	// ChannelMember: expand model.ChannelMember if ChannelID and
+	// ActingUserID are set.
```
Comment (hanzei) [inline apps/expand.go:52]: Is there a difference between `all` and `summary`?

File: apps/expand.go [hunk 2]
```diff
@@ -56,10 +60,13 @@ type Expand struct {
 	Post     ExpandLevel `json:"post,omitempty"`
 	RootPost ExpandLevel `json:"root_post,omitempty"`
 
-	// Team: all for model.team, summary for Id, DisplayName, Name, Description,
+	// Team: all for model.Team, summary for Id, DisplayName, Name, Description,
 	// Email, Type.
 	Team ExpandLevel `json:"team,omitempty"`
 
+	// TeamMember: expand model.TeamMember if TeamID and ActingUserID are set.
+	TeamMember ExpandLevel `json:"team_member,omitempty"`
```
Comment (hanzei) [inline apps/expand.go:68]: Could we ever made a case for a `TeamMember` expand for `User`? Should the team and channel expand for  `ActingUser` make it more clear that that it's about `ActingUser`, by e.g. being named `ActingTeamMember`?

File: server/proxy/expand.go [hunk 3]
```diff
@@ -261,7 +291,12 @@ func stripApp(app apps.App, level apps.ExpandLevel) *apps.App {
 	return nil
 }
 
-func getExpandClient(app apps.App, conf config.Config, mm *pluginapi.Client, in Incoming) (mmclient.Client, error) {
+func (p *Proxy) getExpandClient(app apps.App, in Incoming) (mmclient.Client, error) {
+	if p.expandClientOverride != nil {
+		return p.expandClientOverride, nil
+	}
```
Comment (hanzei) [inline server/proxy/expand.go:297]: If a `pluginapi.Client` is passed to `getExpandClient` this workaround is not longer needed. Why does the `Proxy` need a `expandClientOverride`?

File: server/proxy/expand_test.go [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+		{
+			name: "channel_member",
+			tcs: map[string]TC{
+				"happy": {
+					base: apps.Context{
+						ActingUserID:     userID,
+						UserAgentContext: apps.UserAgentContext{ChannelID: channelID},
+					},
+					expectClientCalls: func(client *mock_mmclient.MockClient) {
+						client.EXPECT().GetChannelMember(channelID, userID).Times(4).Return(channelMember, nil)
```
Comment (hanzei) [inline server/proxy/expand_test.go:64]: Why 4 times? I was expecting 2 times.

**Review by hanzei — APPROVED** [review]
Nice work on the tests :+1: 

I'm wondering they already should be split up in multiple test functions are debugging such a large and nested test can be hard.

File: apps/expand.go (hunk 1 above)
Comment (levb) [reply apps/expand.go:52]: no, it's a small-ish struct

File: server/proxy/expand_test.go (hunk 4 above)
Comment (levb) [reply server/proxy/expand_test.go:64]: Good point. Presently expandContext() loads entities for `expand.XXX != ""` and then strips them. This is wrong. I'll file a ticket and cut another PR to fix them all? https://mattermost.atlassian.net/browse/MM-39935

File: apps/expand.go (hunk 2 above)
Comment (levb) [reply apps/expand.go:68]: Going to merge as is, @hanzei if you feel strongly that "ActingUser" prefix is meaningful, I'll do another PR for that.

**Pull request merged**